ethereum/eth2.0-specs
# add previous and current crosslinks

Somewhat deep change to address some fundamental bugs in how we handle crosslinks in state. Addresses the issues raised in #855 but through a simplified mechanism. Previous and current crosslinks are kept in state to ensure can always check `attestation.previous_crosslink` validity for attestations from previous and current epoch.

Changes:
* remove `latest_crosslinks`
* add `current_crosslinks` and `previous_crosslinks` -- this ensures that if an attestation is from previous or current epoch, the requisite `previous_crosslink` data is available in state to validate
* add `previous_crosslink_root` to `Crosslink` and strictly enforce that crosslinks must form a chain to prevent cases in which data is crosslinked and then overwritten.
* ensure that crosslink rewards are only given to attestations that did or could have formed a crosslink
* remove filter condition in `process_crosslinks` that checks for validity of `previous_crosslink`. This is unnecessary due to it already being checked in `process_crosslink` during block processing.
* add some tests for `process_crosslinks`
